### DISCUSSION TITLE
Help needed with some maintenance tasks

### DISCUSSION THREAD

**Discussion by tkralphs** [coin-or/Cbc#378] [Discussion]
Hi all, many of you have been waiting for a Cbc 3.0 release and I would like to thank you for your patience. There is a draft PR open (#374) that I could use some help in finishing. What's left to do doesn't require any knowledge of Cbc's inner workings. All the details are described in the PR, but in short summary, what remains at a minimum is

- some debugging (there are still probably a few surface-level bugs hanging around that I haven't had time to track down) and
- some thorough testing to make sure there are no performance regressions.

I would also really like to create a minimal `CbcSolver` class, rename `CbcMain1()` and move it into that class, make `CbcMain0()` the default constructor of the new `CbcParameters` class, and finally break up the humongous while loop into some smaller pieces so we can reformat the code. Currently, `clang-reformat` can't even parse the code in `CbcSolver.cpp`. Emacs also gets lost when trying to find matching braces parentheses. This makes maintenance particularly difficult. 

None of this would change any functionality and shouldn't take much time. The heavy lifting is done (there are some nice-to-have things listed in the PR that are a little more difficult, but we could push those off to another release if needed). Let me know if you have some bandwidth to pitch in. The best way to contribute would probably be to create additional PRs based off the `refactor` branch.

**Comment by tosttost** [comment]
I'd like to help but unfortunately building the refactor branch fails on my machine:
```
./coinbrew fetch Cbc@refactor
./coinbrew build Cbc@refactor
```

gives
```
/some/path/Cbc/src/CoinSolve.cpp:22:10: fatal error: CbcDebug.hpp: No such fil                                                                                                                                                                     e or directory
 #include "CbcDebug.hpp"
          ^~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [Makefile:1613: cbc-CoinSolve.o] Error 1
make[1]: *** [Makefile:782: all] Error 2
make: *** [Makefile:457: all-recursive] Error 1

Build failed, see error output above
```

> Some debugging (...)

are there any starting points? 

> Some thorough testing (..)

I can use some of our models for that, but only comparering against the master branch as our somewhat hacky cmake+coinbrew build system does not really work with 2.10. Of course only after the refactor branch compiles.

**Reply by tkralphs** [reply]
Great, thanks!! Your help would be very much appreciated. I will take a look at what you're reporting here. It was building for me locally when I made the last commit.

**Reply by tkralphs** [reply]
Oops looks as though I forgot to push `CbcDebug.hpp` to the repo. It's there now, so you should be able to build. But as a fair warning, I haven't really done much testing at all yet, even very basic stuff. Initially, there will need to be some basic debugging just to even get to the point of running over a test set to look for regressions. I'm planning to get back to this again soonish, but would be happy to have some others pitch in. It shouldn't take any knowledge of the internal workings of Cbc to do the necessary debugging, it's stuff on the level of just ensuring that the command line is parsed correctly, etc.

Also, I noticed that the unit test is not building correctly, which I'll try to fix, but for now, you would build with `--tests none` just to see if the build goes through.

**Reply by tkralphs** [reply]
I rebased the `refactor` branch on `master` to resolve conflicts that were preventing the CI from running, so that we'll now at least know what's happening with the build and test. That was not an easy rebase, as many things changed in master in the meantime, so we should really try to get the refactor branch merged soon.

**Reply by tkralphs** [reply]
I made some obvious fixes and now the unit test is building, but the Gams test is hanging on what is probably another obvious bug in parsing parameters. I'm out of time for now, will try more later.

**Reply by jhmgoossens** [reply]
Is the idea that the cbc code in `Cbc/refactor` branch is built against the `refactor` branches in the related project, .e.g. `Clp/refactor` ? 
I tried that, but `CoinUtils/refactor` is behind its master (such as the fix in commit 6cc871b3 for unary minus), resulting in compile errors. I'm trying to build in VS2019, without using coinbrew.

**Reply by tkralphs** [reply]
Yes, that's the idea. I'll rebase CoinUtils (and also Clp), which should take care of your issues.

**Reply by jhmgoossens** [reply]
Thanks. I tried myself to merge CoinUtils changes from master, but the conflicts in travis and configure I didn't know how to resolve.

**Reply by tkralphs** [reply]
FYI, I just committed some additional bug fixes to Cbc refactor branch. Now, the gamsTest runs successfully at least. There may still be some shallow bugs around, but we're getting closer to having something fully working.

**Reply by tkralphs** [reply]
I rebased one more time to resolve some conflicts with master (and also rebased CoinUtils now). I'll attempt to get all the developers to start working on the refactor branch going forward to avoid further conflicts. I can also switch to merging if anyone is going to start with some serious debugging. I'm going to attempt to set this aside for a while...

**Reply by tkralphs** [reply]
I couldn't resist a little more testing and it looks like there's something fundamentally broken with parsing the command line, so I guess I can't set this aside just yet :).

**Reply by tkralphs** [reply]
OK, just committed a few more small fixes, now it seems to solve instances as expected, at least with simple invocations from the command line. I just noticed, though that valgrind is complaining about conditional branch based on uninitialized memory so a few bugs there to track down.

**Reply by tosttost** [reply]
I already tracked 'conditional branch based on uninitialized memory'  down but the PR needs some cleanup - hopefully I find time to finish it tonight.

**Reply by tkralphs** [reply]
Great!

**Reply by tkralphs** [reply]
I now rebased the Clp refactor branch on master.

**Reply by tkralphs** [reply]
Finally, all unit tests are passing, although there are still some failures in the Travis builds due to symbols not being exported properly.

**Comment by jhmgoossens** [comment]
I looked quite long and hard at the "ProximityNumber" code in CbcSolverHeuristics ("`//TODO: There is a bug here.`"), but couldn't wrap my head around the `fakeKeyword`, `fakeValue`, etc. in the old `currentOptionAsInteger(..)`. Anyway, where have the parameter descriptions of old clp's CbcOrClpParam.cpp moved to?

**Reply by tkralphs** [reply]
The descriptions of parameters can now be found in in `CbcParameters.cpp` and `ClpParameters.cpp` respectively, in the functions that set up the parameters from the constructor of the `CbcParameters` and `ClpParameters` classes. Here
https://github.com/coin-or/Cbc/blob/59145eea450fd7e5555bce0c3c104f670efd3df2/src/CbcParameters.cpp#L1549-L1562) 
is the description of `PROXIMITY`.

**Reply by tkralphs** [reply]
OK, I sort of figured out what's going on there. The parameter `PROXIMITY` has a sort of double meaning, but it never seems to have been fully implemented. It is a keyword parameter, which takes values `off`, `on`, `both`, `before`, like the other heuristics, but it can also take on values `10`, `100`, and `300` (as strings, not integers). So what we really need is a separate parameter to control the `proximityNumber`, but I just left as it was for now, except checking to see whether the value is one of `10`, `100`, and `300` and if so, set `proximityNumber` appropriately.
